### ISSUE TITLE
same `X`, both `X == 0` and `X != 0` are triggered

### ISSUE THREAD

**Issue by loynoir** [rust-lang/rust#145163]
I tried this code:

```rust
    if status == napi_ok {
        eprintln!("[reproduce] status == napi_ok, should not error");
    } else {
        eprintln!("[reproduce] status != napi_ok, should be error");
    }

    if status != napi_ok {
        unsafe {
            napi_throw_error(env, null_mut(), c"[reproduce] Error: status != napi_ok".as_ptr());
        }
        return null_mut();
    }
```

I expected to see this happen: *explanation*

`X == 0` and `X != 0` can not be both triggered

```console
=== Calling debug reproduce ===
[reproduce] status == napi_ok, should not error
```

```console
=== Calling release workaround ===
[workaround] status == napi_ok, should not error
```

Instead, this happened: *explanation*

`X == 0` and `X != 0` can be both triggered

```console
=== Calling release reproduce ===
[reproduce] status == napi_ok, should not error
Error: [reproduce] Error: status != napi_ok
    at Object.<anonymous> (/workspaces/loynoir/repo/reproduce-rust-145162/reproduce.js:9:15)
    at Module._compile (node:internal/modules/cjs/loader:1738:14)
    at Object..js (node:internal/modules/cjs/loader:1871:10)
    at Module.load (node:internal/modules/cjs/loader:1470:32)
    at Module._load (node:internal/modules/cjs/loader:1290:12)
    at TracingChannel.traceSync (node:diagnostics_channel:322:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:238:24)
    at Module.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:154:5)
    at node:internal/main/run_main_module:33:47
```

### Meta

`rustc --version --verbose`:
```
$ rustc --version --verbose
rustc 1.91.0-nightly (ec7c02612 2025-08-05)
binary: rustc
commit-hash: ec7c02612527d185c379900b613311bc1dcbf7dc
commit-date: 2025-08-05
host: x86_64-unknown-linux-gnu
release: 1.91.0-nightly
LLVM version: 20.1.8
```

<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>

## reproduce

https://github.com/loynoir/reproduce-rust-136773

**Comment by loynoir** [comment]
Code updated

## workaround

```rs
let result = MaybeUninit::<napi_value>::uninit();
```

```rs
let result = MaybeUninit::<[napi_value; 1]>::uninit();
```

**Comment by fmease** [comment]
Oh, yeah:

> ```rs
> let args = MaybeUninit::<[napi_value; 2]>::uninit();
> let mut args = unsafe { args.assume_init() };
> ```

**Comment by loynoir** [comment]
> Oh, yeah:
> 
> > let args = MaybeUninit::<[napi_value; 2]>::uninit();
> > let mut args = unsafe { args.assume_init() };

So, what is the recommened rust way, allocate memory, no need for zero, before calling C api get output value?

**Comment by loynoir** [comment]
And what is the general way to lint UB without open an issue?

```sh
cargo miri test
```

miri report nothing.


And I cannot find this kind maybe uninit UB document location.

All I see is 

```rs
MaybeUninit::uninit().assume_init()
```

Above is UB.

**Comment by loynoir** [comment]
Wow, problem fixed!

You guys are really awesome!

And...

Is below code OK or UB?

```rs
    let mut args = MaybeUninit::<[napi_value; 2]>::uninit();

    let status = unsafe { napi_get_cb_info(env, info, argc.as_mut_ptr(), args.as_mut_ptr().cast::<napi_value>(), null_mut(), null_mut()) };

    // should I?
    let _args = unsafe { args.assume_init() };
```

Is above unsafe OK?

**Comment by loynoir** [comment]
If C API somehow did not init maybeuninited mut ptr

- How to prevent my other code got ruined, such as `X == 0` at the same time `X != 0`?

**Comment by loynoir** [comment]
Although problem partially fixed, I take a brief glance at

https://doc.rust-lang.org/reference/behavior-considered-undefined.html

I still no idea which UB rule describe this issue.

**Comment by ChrisDenton** [comment]
See the safety section on [`assume_init`](https://doc.rust-lang.org/stable/std/mem/union.MaybeUninit.html#method.assume_init)

The documentation for [MaybeUninit](https://doc.rust-lang.org/stable/std/mem/union.MaybeUninit.html) has a long discussion about how it should (and shouldn't) be used and why it's UB to use it incorrectly.

**Comment by loynoir** [comment]
I run `cargo miri test`, and got new questions.

I found a dependency UB caused by crate `X` unsafe function `f`

```console
Undefined Behavior: \
`ptr_offset_from` called on two different pointers \
where the memory range between them \
is not in-bounds of an allocation
```

Personally, I currently understand it seems maybe backdoor.

Being afraid of revenged against, I'm not gonna mention it without anonymous.

Brief

- crate `X` unsafe function `f` does mention `MAY` UB when misuse.

- crate `Y` depends on crate `X`

- crate `Y` and crate `X` are from **SAME AUTHOR**

- crate `Y` have millions download

- crate `Y` claim to be safe

- crate `Y` born as crate `Z` deprecated

- crate `Z` have hundred millions download

Question

1. **IMPORTATN** What kind of report does miri makes, describe in style rfc2119?

- `MAY` UB?

- `MUST` UB?

2. **IMPORTATN** What range of UB will miri cover?

- All compile time UB?

- Subset compile time UB?

- All runtime UB?

- Subset compile time UB?

3. Do you guys thinks this seems like backdoor? Should I report crate?

4. Does report crate require gh login?

**Comment by loynoir** [comment]
I run `cargo miri test`, and got new questions.

I found a dependency UB caused by crate `X` unsafe function `f`

```console
Undefined Behavior: \
`ptr_offset_from` called on two different pointers \
where the memory range between them \
is not in-bounds of an allocation
```

Personally, I currently understand it seems maybe backdoor.

Being afraid of revenged against, I'm not gonna mention it without anonymous.

Brief

- crate `X` unsafe function `f` does mention `MAY` UB when misuse.

- crate `Y` depends on crate `X`

- crate `Y` and crate `X` are from **SAME AUTHOR**

- crate `Y` have millions download

- crate `Y` claim to be safe

- crate `Y` born as crate `Z` deprecated

- crate `Z` have hundred millions download

Question

1. **IMPORTANT** What kind of report does miri makes, describe in style `rfc2119`?

- `MAY` UB?

- `MUST` UB?

2. **IMPORTANT** What range of UB will miri cover?

- All compile time UB?

- Subset compile time UB?

- All runtime UB?

- Subset compile time UB?

3. Do you guys thinks this seems like backdoor? Should I report crate?

4. Does report crate require gh login?

**Comment by loynoir** [comment]
Because hundred million download crate `Z` is dead

Thus I depends on million download crate `Y`, which claim safe and fast.

But `Y` depends on **SAME AUTHOR** crate `Z` unsafe function `f`.

Unsafe fn `f` did claim **misuse** leads to UB.

But, mirir **ALWAYS** report UB.

Thus

- consider `Y` and `Z` are from same author, less likely misuse

I have concern about this `ptr_offset_from` might be backdoor.

Would be nice if it isn't. :)

**Comment by workingjubilee** [comment]
loynoir, we cannot help you if you are unwilling to specify more about the author, the exact code, etc. Unsafe code is very subtle, so it is almost useless to talk about it in generalities, without discussing specific instances. I am going to close this issue as it has served its purpose.

Though to answer one thing: miri's reports are not really well-described by either MUST or MAY, nor is UB well-described by either MUST or MAY, because neither is based on an understanding of `unsafe fn` which are source-level entities. They are purely based on the operation of the code.
